SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialAuthorityTest/SpatialAuthorityTest.cpp

```diff
@@ -0,0 +1,387 @@
+  *		- dynamic non-replicated actor
+  *		- GameMode (which only exists on Servers)
+  *		- GameState
+  * Keep in mind that we're assuming a 1x2 Grid Load-Balancing Strategy, otherwise the ownership of
+  * these actors may be something completely different (specially important for actors placed in the Level).
+  * You have some flexibility to change the Server1/2Position properties to test in different Load-Balancing Strategies.
+  */
+ASpatialAuthorityTest::ASpatialAuthorityTest()
+{
+	Author = "Nuno Afonso";
+	Description = TEXT("Test HasAuthority under multi-worker setups. It also ensures it works in Native");
+
+	Server1Position = FVector(-250.0f, -250.0f, 0.0f);
+	Server2Position = FVector(-250.0f, 250.0f, 0.0f);
+}
+
+void ASpatialAuthorityTest::BeginPlay()
+{
+	Super::BeginPlay();
+
+	ResetTimer();
+
+	// Replicated Level Actor. Server 1 should have Authority, again assuming that the Level is setup accordingly.
+	{
+		AddStep(TEXT("Replicated Level Actor - Server 1 Has Authority"), FWorkerDefinition::AllWorkers,	nullptr, nullptr,
+			[this](float DeltaTime) {
+				Timer -= DeltaTime;
+				if (Timer <= 0)
+				{
+					const FWorkerDefinition& LocalWorkerDefinition = GetLocalFlowController()->WorkerDefinition;
+					if (LocalWorkerDefinition.Type == ESpatialFunctionalTestWorkerType::Server && LocalWorkerDefinition.Id == 1)
+					{
+						if (VerifyTestActor(LevelReplicatedActor, 1, 1, 1, 0))
+						{
+							FinishStep();
+						}
+					}
+					else
+					{
+					
+						if (VerifyTestActor(LevelReplicatedActor, 0, 0, 0, 0))
+						{
+							FinishStep();
+						}
+					}
+				}
+			}, 5.0f);
+	}
+
+	// Non-replicated Level Actor. Each Server should have Authority over their instance, Clients don't.
+	{
+		AddStep(TEXT("Non-replicated Level Actor - Each Server has Authority, Client doesn't know"), FWorkerDefinition::AllWorkers, nullptr, nullptr,
+			[this](float DeltaTime) {
+				// Since this actor already was in level and we wait for timer in the previous step, we don't need to wait
+				// in this one again.
+				const FWorkerDefinition& LocalWorkerDefinition = GetLocalFlowController()->WorkerDefinition;
+				if (LocalWorkerDefinition.Type == ESpatialFunctionalTestWorkerType::Server)
+				{
+					// Note: Non-replicated actors never get OnAuthorityGained() called.
+					if (VerifyTestActor(LevelActor, LocalWorkerDefinition.Id, LocalWorkerDefinition.Id, 0, 0))
+					{
+						FinishStep();
+					}
+				}
+				else
+				{
+					if (VerifyTestActor(LevelActor, 0, 0, 0, 0))
+					{
+						FinishStep(); // Clients don't have authority over non-replicated Level Actors.
+					}
+				}
+			}, 5.0f);
+	}
+
+	// Replicated Dynamic Actor Spawned On Same Server. Server 1 should have Authority.
+	{
+		AddStep(TEXT("Replicated Dynamic Actor Spawned On Same Server - Spawn"), FWorkerDefinition::Server(1), nullptr,
+			[this]()
+			{
+				ASpatialAuthorityTestReplicatedActor* Actor = GetWorld()->SpawnActor<ASpatialAuthorityTestReplicatedActor>(
```

Comment:
How does server 1 set DynamicReplicatedActor ?

Reply:
Oh I see it's replicated, and the cross server call just ensures it's set first on the correct server

Reply:
Do we statically define which server should be authoritative over ASpatialAuthorityTest ?

Reply:
No, Tests will be owned by the worker where they are placed currently.